saml/src/main/java/com/linecorp/armeria/server/saml/HttpRedirectBindingUtil.java

```diff
@@ -0,0 +1,341 @@
+     */
+    static void validateSignature(Credential validationCredential,
+                                  SamlParameters parameters,
+                                  String messageParamName) throws SamlException {
+        requireNonNull(validationCredential, "validationCredential");
+        requireNonNull(parameters, "parameters");
+        requireNonNull(messageParamName, "messageParamName");
+
+        final String signature = parameters.getFirstValue(SIGNATURE);
+        final String sigAlg = parameters.getFirstValue(SIGNATURE_ALGORITHM);
+
+        // The order is one of the followings:
+        // - SAMLRequest={value}&RelayState={value}=SigAlg={value}
+        // - SAMLResponse={value}&RelayState={value}=SigAlg={value}
+        final QueryStringBuilder builder = new QueryStringBuilder();
+        builder.addParam(messageParamName, parameters.getFirstValue(messageParamName));
+
+        final String relayState = parameters.getFirstValueOrNull(RELAY_STATE);
+        if (relayState != null) {
+            builder.addParam(RELAY_STATE, relayState);
+        }
+        builder.addParam(SIGNATURE_ALGORITHM, sigAlg);
+
+        final byte[] input = builder.toString().getBytes(StandardCharsets.UTF_8);
+
+        try {
+            final byte[] decodedSignature = Base64.getDecoder().decode(signature);
+            if (!XMLSigningUtil.verifyWithURI(validationCredential, sigAlg, decodedSignature, input)) {
+                throw new SamlException("failed to validate a signature");
+            }
+        } catch (IllegalArgumentException e) {
+            throw new SamlException("failed to decode a base64 signature string", e);
+        } catch (SecurityException e) {
+            throw new SamlException("failed to validate a signature", e);
+        }
+    }
+
+    /**
+     * Generates a signature of the specified {@code input}.
+     */
+    @VisibleForTesting
+    static String generateSignature(Credential signingCredential,
+                                    String algorithmURI, String input) throws SamlException {
+        try {
+            final byte[] signature =
+                    XMLSigningUtil.signWithURI(signingCredential, algorithmURI,
+                                               input.getBytes(StandardCharsets.UTF_8));
+            return Base64.getEncoder().encodeToString(signature);
+        } catch (SecurityException e) {
+            throw new SamlException("failed to generate a signature", e);
+        }
+    }
+
+    /**
+     * Encodes the specified {@code message} into a deflated base64 string.
+     */
+    static String toDeflatedBase64(SAMLObject message) throws SamlException {
+        requireNonNull(message, "message");
+
+        final String messageStr;
+        try {
+            messageStr = nodeToString(XMLObjectSupport.marshall(message));
+        } catch (MarshallingException e) {
+            throw new SamlException("failed to marshal a SAML message", e);
```

Comment:
I'm curious which is correct, marshal or marshall? Alternatively, we could just use 'serialize' and 'deserialize'. :-D

Reply:
Fixed using 'serialize' and 'deserialize'. Thanks!